### ISSUE TITLE
Using a MirroredStrategy with an ImageDataGenerator does not work correctly

### ISSUE THREAD

**Issue by haimat** [keras-team/keras#15692]
**System information**.
- Have I written custom code (as opposed to using a stock example script provided in Keras): yes
- OS Platform and Distribution (e.g., Linux Ubuntu 16.04): Ubuntu 20.04
- TensorFlow installed from (source or binary): via python pip
- TensorFlow version (use command below): v2.7.0-rc1-69-gc256c071bb2
- Python version: 3.8.10
- GPU model and memory: NVIDIA GeForce RTX 2070
- Exact command to reproduce: just run the sample code provided below

**Describe the problem**.

I am training a custom CNN model in Keras, based on pre-made models from `tensorflow.keras.applications` (e.g. VGG16). Using the *Kaggle Cats vs. Dogs* dataset I cam easily achieve an accuracy of 97% and more. In order to read-in the images to train my model, I am using the Keras `ImageDataGenerator` class. I want to make use of it since it allows me to easily a) augment images on the fly and b) apply a pre-processing function to each input image.

Now I want to extend my script to allow for multi-GPU training. So I am following the official [Keras Multi-GPU training docs](https://keras.io/guides/distributed_training/) in order to implement single-host, multi-device synchronous training. That's easy enough - I simply created a `MirroredStrategy` object, then build and compile my Keras model within the strategy scope - everything as described in the docs.

**Describe the current behavior**.

When training starts via `.fit()` TensorFlow tells me that it can't apply the FILE sharding policy:

> AUTO sharding policy will apply DATA sharding policy as it failed to apply FILE sharding policy because of the following reason: Did not find a shardable source, walked to a node which is not a dataset

This seems to be an error, since I am using thousands of image files per class, thus it should easily be possible to split the input data into batches for multiple GPUs based on these files.

**Describe the expected behavior**.

I don't unerstand where this error comes from. However, to be sure I followed the advice given by TensorFlow, to set the sharding policy to DATA:

```python
options = tfd.Options()
options.experimental_distribute.auto_shard_policy = tfd.experimental.AutoShardPolicy.DATA
training_dataset.with_options(options)
```

But that doesn't change a bit - I still get exactly the same error message.  
Which leads me to two questions:

1. What do I need to do in order to get rid of this warning?
2. **Even more important**: Why is Keras not able to split the dataset with the default `AutoShardPolicy.FILE` policy, since I am using thousands of images per class in the input folder?

**Standalone code to reproduce the issue**.
```python
from tensorflow.keras.preprocessing.image import ImageDataGenerator
from tensorflow.keras.applications import VGG16
from tensorflow.keras.applications import vgg16
from tensorflow.keras import models
from tensorflow.keras import layers
import tensorflow as tf

# Download Kaggle Cats vs. Dogs - https://www.kaggle.com/c/dogs-vs-cats/data?select=train.zip
TRAIN_DS_PATH = "/tmp/catsvsdogs"
IMAGE_CLASSES = ["cat", "dog"]
BATCH_SIZE = 32
IMG_SIZE = 224
EPOCHS = 10

def build_model():
    input_layer = layers.Input(shape=(IMG_SIZE, IMG_SIZE, 3), name="model_input")
    base_model = VGG16(weights="imagenet", include_top=False, input_tensor=input_layer)
    base_model.trainable = False
    model_head = base_model.output
    model_head = layers.AveragePooling2D(name="custom_header", pool_size=(4, 4))(model_head)
    model_head = layers.Flatten()(model_head)
    model_head = layers.Dense(256, activation="relu")(model_head)
    model_head = layers.Dropout(0.5)(model_head)
    model_head = layers.Dense(len(IMAGE_CLASSES), activation="softmax")(model_head)
    return models.Model(inputs=input_layer, outputs=model_head)

# Create the image data generator for our trainig data
image_datagen = ImageDataGenerator(preprocessing_function=vgg16.preprocess_input)
training_img_generator = image_datagen.flow_from_directory(
    directory=TRAIN_DS_PATH,
    target_size=(IMG_SIZE, IMG_SIZE),
    batch_size=BATCH_SIZE,
    class_mode="categorical",
)
steps = training_img_generator.n // training_img_generator.batch_size

# Create a TF dataset based on the Keras image data generator
training_dataset = tfd.Dataset.from_generator(
    lambda: training_img_generator,
    output_types=(tf.float32, tf.float32),
    output_shapes=([None, IMG_SIZE, IMG_SIZE, 3], [None, len(IMAGE_CLASSES)]),
)

# Set the sharding policy to DATA
options = tf.data.Options()
options.experimental_distribute.auto_shard_policy = tf.data.experimental.AutoShardPolicy.DATA
training_dataset.with_options(options)

# Build and compile the Keras model within a mirrored strategy
strategy = tf.distribute.MirroredStrategy()
with strategy.scope():
    keras_model = build_model()
    keras_model.compile(loss="binary_crossentropy", optimizer="adam", metrics=["accuracy"])

# Last but not least train the model via .fit()
keras_model.fit(training_dataset, steps_per_epoch=steps, epochs=EPOCHS)
```

**Comment by haimat** [comment]
Hello, I noticed this issue is not marked as bug any more - which suprises me, because I had the impression that since I am using two "out of the box" parts from TF & Keras and following the official docs, that should work. However, if this is not a bug, could you please be so kind and let me know how else I should proceed, what other approach I would need to take? Thanks in advance!

**Comment by haimat** [comment]
Update: I tried it now with `from_tensor_slices()`, like so:

```python
@tf.function
def load_images(image_path, label):
    image = tf.io.read_file(image_path)
    image = tf.image.decode_jpeg(image, channels=3)
    image = vgg16.preprocess_input(image)
    image = tf.image.resize(image, (IMG_SIZE, IMG_SIZE))
    return (image, label)

training_dataset = tf.data.Dataset.from_tensor_slices((train_paths, train_labels_le))
training_dataset = (
    training_dataset.shuffle(1024)
    .map(load_images, num_parallel_calls=tf.data.AUTOTUNE)
    .batch(BATCH_SIZE)
    # .map(lambda x, y: (data_augmentation(x), y), num_parallel_calls=tf.data.AUTOTUNE)
    .prefetch(tf.data.AUTOTUNE)
)
```

However, evem though model training is fine, this leads to exactly the same warning message:

> AUTO sharding policy will apply DATA sharding policy as it failed to apply FILE sharding policy because of the following reason: Found an unshardable source dataset: name: "TensorSliceDataset/_2"

**So how is one supposed to create and use a TF Dataset for multi-GPU training in Keras?**

**Comment by sachinprasadhs** [comment]
You can refer to [this](https://www.tensorflow.org/tutorials/distribute/input#sharding) document on distributed input for details on data sharding. Let me know if this helps you. Thanks!

**Comment by haimat** [comment]
Thanks for the link. Yesterday I seem to got it working via switching to a `TFRecordDataset`.
When working with TFRecords instead of image data generators (but not other changes), then it seems to work!

**Comment by sachinprasadhs** [comment]
Could you please move this issue to closed, if it is resolved for you. Thanks!

**Comment by varungupta31** [comment]
Typically, we would use `ImageDataGenerator`, and apply desired pre-processing and augmentations, then say, the `flowfromdirectory` method to load the data with corresponding labels in desired (one-hot, binary, etc.) format.

Now, if I were to use distributed GPU training, as @haimat mentioned, how will I apply these operations (or basically mimic this workflow) with `TFRecordDataset`?

Say, for this type of operation over my data and then use it for distributed GPU training:
```
datagen = tf.keras.preprocessing.image.ImageDataGenerator(
    preprocessing_function=tf.keras.applications.vgg19.preprocess_input,
    rescale=1.0/255.0,
    horizontal_flip=True,
    zoom_range=0.2,
     shear_range=0.2)
```
